### PR TITLE
wallet: Remove boost from PeriodicFlush

### PR DESCRIPTION
**Pull request by maflcko**
The `boost::this_thread::interruption_point()` in the code base currently block the replacement of `boost::thread` with `std::thread`. [1]

Remove them from the wallet because they are either unused or useless.

The feature to interrupt a periodic flush is useless because all wallets have just been flushed https://github.com/bitcoin/bitcoin/blob/9ccaee1d5e2e4b79b0a7c29aadb41b97e4741332/src/init.cpp#L194 and another flush should be a noop. Also, they will be flushed again shortly after https://github.com/bitcoin/bitcoin/blob/9ccaee1d5e2e4b79b0a7c29aadb41b97e4741332/src/init.cpp#L285, so even if repeated flushes weren't a noop, doing 3 instead of 2 shouldn't matter too much at this point. Also, the wallet is flushed every two seconds in the worst case, so if this is an expensive operation, that period should be readjusted. (Or bdb should be removed altogether #18916)

[1] Replacement of `boost::thread` with `std::thread` should happen because:

* The boost thread dependency is slow to compile
* Boost thread is less maintained than the standard lib
* Boost thread is mostly redundant to the standard lib
* Global interruption points via exceptions are hard to keep track of during review and easy to get wrong during runtime (e.g. accidental `catch (...)`)

### CODE REVIEW

**Comment by MarcoFalke** [comment]
Not sure how to test manually, but I have included a unit test

**Comment by hebasto** [comment]
Concept ACK.

**Comment by practicalswift** [comment]
Concept ACK

**Comment by laanwj** [comment]
The main thing to review here is the change in behavior: `interruption_point` raises an exception that escalates all the way to the thread main level, while the new code returns `false`, which is handled differently. For example it will still continue iterating over `for (const std::shared_ptr<CWallet>& pwallet : GetWallets()) {`. so `dbh.nLastSeen` and `dbh.nLastWalletUpdate` will be updated for all of them. I do not know if this is a problem.

**Comment by MarcoFalke** [comment]
> I do not know if this is a problem

Me neither, but it seems avoidable by adding one more if-check. Done that.

**Comment by laanwj** [comment]
Shouldn't the check be at the beginning of the for loop instead of of `MaybeCompactWalletDB` in that case?

**Comment by MarcoFalke** [comment]
Thanks, fixed.

**Comment by MarcoFalke** [comment]
Rebased after silent merge conflict

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -433,6 +433,8 @@ void CWallet::Flush(bool shutdown)
     database->Flush(shutdown);
 }
 
+std::atomic<bool> CWallet::m_end_flush{false};
```
Comment (jnewbery) [inline src/wallet/wallet.cpp:436]: Why not use default member initialization in the header file?

**Review by jnewbery — COMMENTED** [review]
Concept ACK. Can you add a description for this PR?

You say here: https://github.com/bitcoin/bitcoin/pull/18758#issue-408624391 that these interruption points can't be used. Were they used at some point in the past and were broken? If so, have you been able to determine which commit broke them? If not, I don't think it's worth adding the extra shutdown handling in this PR. It could potentially be added in a future PR.

You can remove the `#include <boost/thread.hpp>` from `walletdb.cpp` and `db.cpp`.

File: src/wallet/wallet.cpp (hunk 1 above)
Comment (MarcoFalke) [reply src/wallet/wallet.cpp:436]: This wouldn't compile. Thanks C++

**Comment by MarcoFalke** [comment]
> Concept ACK. Can you add a description for this PR?

Done

**Comment by MarcoFalke** [comment]
> You say here: #18758 (comment) that these interruption points can't be used.

Sorry, that pull is still marked as draft and is not ready for review. The decision boils down to "Is the interruption point run in a `std::thread`? If yes, the interruption point is unused".

*Here*, the scheduler is (still) a `boost::thread`, so the point should not be unused.

**Comment by MarcoFalke** [comment]
Force pushed to address feedback by @jnewbery (remove boost include from db.cpp)

**Comment by jnewbery** [comment]
> Force pushed to address feedback by @jnewbery (remove boost include from db.cpp)

I also suggested removing it from `walletdb.cpp`. Is that not possible?

**Comment by jnewbery** [comment]
> the scheduler is (still) a boost::thread, so the point should not be unused.

From a quick read of https://www.boost.org/doc/libs/1_53_0/doc/html/thread/thread_management.html#thread.thread_management.tutorial.interruption, it seems to me that the `boost::thread::interruption_point` only does anything if that thread's `interrupt()` function is called.

The only place that happens is the `threadGroup.interrupt_all()` call in `Shutdown()`, which (since #18234) happens after the scheduler thread is stopped.

I haven't looked deeply enough into #18234 to figure out whether that's a regression. If not, I think we can just remove all of these interrupt points that no longer do anything. If it is a regression, then perhaps this PR should explicitly state that it's a bug fix.

**Comment by MarcoFalke** [comment]
walletdb.cpp is still using it https://github.com/bitcoin/bitcoin/blob/fabd1c91b4c715f0b0cfa28457cf766246568449/src/wallet/walletdb.cpp#L742

**Comment by MarcoFalke** [comment]
> The only place that happens is the threadGroup.interrupt_all() call in Shutdown(), which (since #18234) happens after the scheduler thread is stopped.

It looks like 306f71b4eb4a0fd8e64f47dc008bc235b80b13d9 tells the scheduler to stop *after* the current task is executed or *before* the execution starts (i.e. the scheduler is waiting for the next flush). The call is non-blocking when a task is currently executed. Consequently, the next line will interrupt all threads in the thread group, which will also interrupt the periodic flush action in case it is currently running. So I don't think the behaviour has changed in that commit, at least not in regard to flushing? Correct me if I am wrong.

**Comment by MarcoFalke** [comment]
I start to wonder if it is worth to abort the flush when it is called again shortly after in `StopWallets`.

**Comment by jnewbery** [comment]
> walletdb.cpp is still using it ...

That's a catch in `WalletBatch::LoadWallet()`, which I think is only called in the main thread or rpc thread.

**Comment by jnewbery** [comment]
> It looks like 306f71b tells the scheduler to stop after the current task is executed or before the execution starts (i.e. the scheduler is waiting for the next flush). The call is non-blocking when a task is currently executed. Consequently, the next line will interrupt all threads in the thread group, which will also interrupt the periodic flush action in case it is currently running. So I don't think the behaviour has changed in that commit, at least not in regard to flushing? Correct me if I am wrong.

No, you're right. I misunderstood what `stop()` was doing.

> I start to wonder if it is worth to abort the periodic flush when flush(true) is called shortly after in StopWallets.

Yes, it does seem unnecessary. We call `flush()` on all wallets _before_ interrupting the scheduler thread:

https://github.com/bitcoin/bitcoin/blob/master/src/init.cpp#L194

then we interrupt the scheduler thread (which may cause us to interrupt a periodic flush), and then we flush again on shutdown:

https://github.com/bitcoin/bitcoin/blob/master/src/init.cpp#L285

Making that middle flush interruptible does seem unnecessary.

**Comment by MarcoFalke** [comment]
Simplified pull request to only remove code :rocket:

**Comment by jnewbery** [comment]
concept ACK just removing the boost interruption stuff. It's essentially unused for the reason you give. We'd only ever interrupt a flush that we're doing between two other flushes.

utACK fa7b885f51ff848d3f913bc6e15d24528300c210

**Comment by MarcoFalke** [comment]
@laanwj @hebasto Mind taking another look here?

**Review by fanquake — CHANGES_REQUESTED** [review]
ACK the change - however there's another `boost::thread_interrupted&` in walletdb.cpp:

https://github.com/bitcoin/bitcoin/blob/44307449f758b239fca0ad6b722bc36795ac6727/src/wallet/walletdb.cpp#L890-L893

I think it can also be removed, but if not, I'm not sure how we can drop the `<boost/thread.hpp>` include as unused in fa7b885f51ff848d3f913bc6e15d24528300c210.

**Comment by MarcoFalke** [comment]
@fanquake Good catch. Added a commit with rationale to remove that as well.

**Review by fanquake — APPROVED** [review]
ACK fa1c74fd0342b74d44cc4e41fff3890c1434e8f7

**Pull request merged**